### PR TITLE
feat: Add option to skip heathcheck on exposed ports.

### LINKED ISSUE #6160: Skip health check on ExposedPort
### What are you trying to do?

I'm running couchbase as a service. It need to expose all the 6+ ports of couchbase so it can be accessed properly. But only the 8091 should be checked for health, because the other ports are unreliable. I'm using the Go SDK.

### Why is this important to you?

I would like to have an option in the `ContainerWithExposedPortOpts` so I can disable the health check for certain ports.

### How are you currently working around this?

I don't have a workaround, because I need to expose the ports, but the execution hangs, because the ports can not be checked.

### PR DESCRIPTION
**Pull request by KGB33**
Closes #6160. 

I'm not sure where, if at all, the best place to test this would be. My best guess is `core/services_test.go`.

Happy to change the naming, or add documentation/cookbook examples if needed. 

~~Changie is WIP, I need the PR number from this.~~

### CODE REVIEW

**Comment by gerhard** [comment]
WDYT @helderco @vito ?

File: core/healthcheck.go [hunk 1]
```diff
@@ -34,7 +34,9 @@ func (d *portHealthChecker) Check(ctx context.Context) (err error) {
 
 	args := []string{"check", d.host}
 	for _, port := range d.ports {
-		args = append(args, fmt.Sprintf("%d/%s", port.Port, port.Protocol.Network()))
+		if !port.SkipHealthCheck {
+			args = append(args, fmt.Sprintf("%d/%s", port.Port, port.Protocol.Network()))
+		}
```
Comment (vito) [inline core/healthcheck.go:39]: Seems like we'll need to handle the case where all ports are `SkipHealthCheck`. If that's the case we should just skip the whole process. I'm not sure if it does, but I could imagine the `check` command failing if it isn't given any ports to check. (This probably warrants an integration test of its own.)

**Review by vito — COMMENTED** [review]
Thanks for taking the initiative here! The feature makes sense. To answer your question, a test should be added to `core/integration/services_test.go` - let us know if you need any help with that.

API-wise I can think of things to bikeshed, but I don't want to expand scope needlessly and leave this blocked, so I'd be good with just marking this "experimental" for now and moving on (by renaming it to `experimentalSkipHealthcheck` - quite a mouthful).

But if there's appetite for bikeshedding, configurable healthchecks is what I have in mind. I can think of a few that folks might want someday:

* HTTP (`GET /foo` with timeout, expect 2xx)
* gRPC? (including for completeness, I see K8s supports it)
* Watch for a string in stdout (lots of processes print something when they're "ready")

GIven that, it might make sense to express this as a "None" health-check or something and move the current behavior to a default (probably called "TCP", and we should probably stop health-checking UDP ports since that's nonsense, we should just check that the hostname resolves).

The last option probably implies a different API since it's process-wide and not related to a particular port.

**Comment by KGB33** [comment]
Configurable health checks sound great!

My first thought is to create a new `Service.with_xyz_healthcheck(...)` API.
It's where I first looked to see if skipping health checks was possible, and I
think it would work better than cramming a bunch of options into
`Container.with_exposed_port(...)`. Plus, moving health checks to `Service`
would prevent confusion if users assumed that health checks configured on
containers in Dagger carried through to exported/published/downloaded
artifacts.

For example:

```python
@function
def mariadb() -> dagger.Service:
    return (
        dagger.container()
        .from_("mariadb:10.11.2")
        .as_service()
        .with_tcp_healthcheck(3306)
    )
```

I'm happy to make a GitHub issue or a thread in discord if you'd like a
dedicated place to bikeshed this. 

Either way I'll look into adding tests to this PR over the weekend.

**Comment by gerhard** [comment]
@vito I think this is related https://github.com/dagger/dagger/issues/5472#issuecomment-1855872373

**Comment by KGB33** [comment]
I renamed the option to `experimentalSkipHealthcheck` for this PR.

I'd still love to implement configurable health checks, and I have a almost-working implementation of my `Service.with_tcp_healthcheck(...)` API [here](https://github.com/KGB33/dagger/pull/1). I want to get more feedback on the API before I put in too much time though.

**Review by jedevc — APPROVED** [review]
> I renamed the option to `experimentalSkipHealthcheck` for this PR.

SGTM :tada: I also have a use case for this, so definitely on the side of getting *something* for this in, and iterating on it later.

I think it would be good to bikeshed the configurable healthchecks in a separate issue, so we can avoid blocking on this.

WDYT @vito?

**Review by vito — APPROVED** [review]
Sorry for the delay, this looks good to me now but just needs a rebase on top of the new DagQL architecture changes. Shouldn't be anything major, but let me know if you have any questions!

File: core/schema/container.go [hunk 2]
```diff
@@ -399,7 +399,8 @@ func (s *containerSchema) Install() {
 				`- For setting the EXPOSE OCI field when publishing the container`).
 			ArgDoc("port", `Port number to expose`).
 			ArgDoc("protocol", `Transport layer network protocol`).
-			ArgDoc("description", `Optional port description`),
+			ArgDoc("description", `Optional port description`).
+			ArgDoc("experimentalSkipHealthCheck", `Skip the health check when run as a service.`),
```
Comment (KGB33) [inline core/schema/container.go:403]: I feel like I'm missing something super simple here... 

If I comment out the ```ArgDoc("experimentalSkipHealthCheck", `Skip the health check when run as a service.`),``` line the SDKs generate & tests pass as expected, But with it seems like the GraphQL server can't startup due to the following error:

```
Error: input:1: pipeline.pipeline.container.from.withExec.withEnvVariable.withExec.withWorkdir.withDirectory.withMountedCache.withExec.withMountedDirectory.withMountedCache.withExec.file process "go run ./cmd/introspect" did not complete successfully: exit code: 1

Stdout:

Stderr:
panic: field withExposedPort has no such argument: "experimentalSkipHealthCheck"

goroutine 1 [running, locked to thread]:
github.com/dagger/dagger/dagql.Field[...].ArgDoc(_, {_, _}, {_, _, _})
        /app/dagql/objects.go:551 +0x274
github.com/dagger/dagger/core/schema.(*containerSchema).Install(0xc000070500)
        /app/core/schema/container.go:403 +0x7e5d
github.com/dagger/dagger/core/schema.(*CoreMod).Install(0x15014a0?, {0xc0000d61e0?, 0x40fade?}, 0xc0000a3b30)
        /app/core/schema/coremod.go:49 +0x4a2
github.com/dagger/dagger/core/schema.New({0x1508a00, 0x1e3b0c0}, {0x0, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}, ...}, ...})
        /app/core/schema/schema.go:65 +0x3f3
main.main()
        /app/cmd/introspect/main.go:12 +0x4e
exit status 2
exit status 1
```

Comment (vito) [reply core/schema/container.go:403]: Looks like the field is `ExperimentalSkipHealthcheck` - the `c` in `check` needs to be capitalized for it to match (or vice versa)

Comment (KGB33) [reply core/schema/container.go:403]: `"experimentalSkipHealthCheck"` vs `"experimentalSkipHealthcheck"` - Capitalized `C`.

**Comment by vito** [comment]
Thanks! :pray:

**Comment by bupd** [comment]
thanks for this.! :+1:

**Pull request merged**